### DISCUSSION TITLE
Iterating all indices of wavemaps

### DISCUSSION THREAD

**Discussion by PaulKemppi** [ethz-asl/wavemap#78] [Q&A] [answered]
Hi and thank you for the extremely interesting library. I have managed to integrate wavemap to my data collection pipeline and I save the created maps to .wvmp files. I would like to compare two created maps collected from the same environment (using the same world coordinate frame). How to iterate all occupied voxels from map1, get the voxel's xyz coordinates and use them to check if the corresponding location of map2 is occupied (or free or unknown). The examples presented for the python / C++ APi are quite minimal and only demonstrate how to get values for single / random set of indices. Perhaps this is a bit noob question, but I think it would be a nice addition to the examples to at least show, how to iterate through all the stored indices of a map.

Best Regards,
Paul

**Comment by victorreijgwart** [comment, marked as answer]
Hi @PaulKemppi , 

Thanks for the good question! To do what you described, you could use the map interface's `forEachLeaf` method. It takes a lambda function, which it then calls for each leaf node in the map - passing the node's index and value as arguments. We didn't get to add it to the Python API yet, though, so it's currently only available in C++. 

To compare `map_a` to `map_b`, you could
* Instantiate a QueryAccelerator for map_b.
* Define a comparison lambda `auto comparator_fn = ...` that captures `accelerator_b` by reference and expects `node_index` and `node_value_a` as arguments. In the lambda, it then uses `accelerator_b` to look up `node_value_b`.
* Dispatch it by calling `map_a.forEachLeaf(comparator_fn)`. 

Note that in some parts of the environment, map_b could have leaf cells at a higher resolution than map_a (especially in areas that changed). However, `map_a.forEachLeaf` method only considers map_a's leaves. One easy workaround is to always process occupied regions at the highest resolution. The code for this would look very similar to what we use to convert wavemaps to fixed resolution point clouds [here](https://github.com/ethz-asl/wavemap/blob/main/interfaces%2Fros1%2Fwavemap_ros%2Fsrc%2Fmap_operations%2Fpublish_pointcloud_operation.cc#L67-L91). 

I hope this helps! Let me know if you have follow-up questions.

**Reply by PaulKemppi** [reply]
Thank you for the quick reply. I think the forEachLeaf  method would be a nice addition to Python API. As I am working with rospy, I ended up subscribing to the occupied voxels pointcloud topic of map1 and then querying the voxels from the map2 using the location of the points. This approach is fast enough at this point. Thanks for the hint. As a result, I can use the comparison (to both directions) in volumetric change detection.
